src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+  private ConcurrentHashMap<String, SinkRecord> recordMap;
+  private ElasticsearchSinkConnectorConfig config;
+  private ErrantRecordReporter reporter;
+  private RestHighLevelClient client;
+
+  public ElasticsearchClient(
+      ElasticsearchSinkConnectorConfig config,
+      ErrantRecordReporter reporter
+  ) {
+    int pos = 0;
+    HttpHost[] hosts = new HttpHost[config.connectionUrls().size()];
+    for (String address : config.connectionUrls()) {
+      try {
+        URL url = new URL(address);
+        hosts[pos++] = new HttpHost(url.getHost(), url.getPort(), url.getProtocol());
+      } catch (MalformedURLException e) {
+        throw new ConnectException(e);
+      }
+    }
+
+    this.error = new AtomicReference<>();
+    this.recordMap = reporter != null || !config.ignoreKey()
+        ? new ConcurrentHashMap<>()
+        : null;
+    this.config = config;
+    this.reporter = reporter;
+    this.client = new RestHighLevelClient(
+        RestClient
+            .builder(hosts)
+            .setHttpClientConfigCallback(this::customizeHttpClientConfig)
+            .setRequestConfigCallback(this::customizeRequestConfig)
+    );
+    this.bulkProcessor = BulkProcessor
+        .builder((req, lis) -> client.bulkAsync(req, RequestOptions.DEFAULT, lis), buildListener())
+        .setBulkActions(config.batchSize())
+        .setConcurrentRequests(config.maxInFlightRequests() - 1) // 0 = no concurrent requests
+        .setFlushInterval(TimeValue.timeValueMillis(config.lingerMs()))
+        .setBackoffPolicy(
+            BackoffPolicy.exponentialBackoff(
+                TimeValue.timeValueMillis(config.retryBackoffMs()),
+                config.maxRetries()
+            )
+        )
+        .build();
+  }
+
+  /**
+   * Returns the underlying Elasticsearch client.
+   *
+   * @return the underlying RestHighLevelClient
+   */
+  public RestHighLevelClient client() {
+    return client;
+  }
+
+  /**
+   * Closes the ElasticsearchClient.
+   *
+   * @throws ConnectException if all of the records fail to flush before the timeout.
+   */
+  public void close() {
+    try {
+      if (!bulkProcessor.awaitClose(1, TimeUnit.MINUTES)) {
+        throw new ConnectException("Failed to process all outstanding requests in time.");
+      }
+    } catch (InterruptedException e) {
+      log.error("Bulk processor close was interrupted.", e);
+    }
+  }
+
+  /**
+   * Creates an index. Will not recreate the index if it already exists.
+   *
+   * @param index the index to create
+   */
+  public void createIndex(String index) {
+    if (indexExists(index)) {
+      return;
+    }
+
+    CreateIndexRequest request = new CreateIndexRequest(index);
+    callWithRetries(
+        "create index " + index,
+        () -> {
+          try {
+            client.indices().create(request, RequestOptions.DEFAULT);
+          } catch (ElasticsearchStatusException | IOException e) {
+            if (!e.getMessage().contains(RESOURCE_ALREADY_EXISTS_EXCEPTION)) {
```

Comment:
Is it still possible to get this exception if we already checked whether the index exists earlier?

Reply:
we have seen it before in escalations and this maintains the previous behavior of the `JestClient` so I kept it just to be safe

unfortunately we have not been able to understand why this happens